src/absil/ilwrite.fsi

```diff
@@ -15,17 +15,22 @@ type ILStrongNameSigner =
     static member OpenKeyPairFile: string -> ILStrongNameSigner
     static member OpenKeyContainer: string -> ILStrongNameSigner
 
+type EmitStreamProvider = Lazy<Choice<EmitTo, Diagnostic>>
```

Comment:
I'd like to encode contract of  the behaviour in the type system.

for me `(unit -> Stream)` is a problem in the contract api for the Stream:
- can be consumed multiple times, more difficult to call the api (need to dispose multiple streams)
- each call can return a different Stream
- i dont know if it'is already called ( i can track, but it'is easier to already know with `Lazy.IsValueCreated` ).

`Lazy` here is about "do that deferred, but exactly once", i can encode this requirement with a type and a DU, but lazy as type already exists.

About the other improvement, the type

```fsharp
| File of string
| Stream of (unit -> Choice<Stream, Diagnostic>)
```

seem better because i dont expected the string filename to return an error.
An api where the type system enforce that an error is not possible is better. :+1:  I'll use your version

Another way of postpone the problem is to check that at higher level and pass *internally* a non lazy 

```
type EmitTo = File of string | Stream of Stream
```
to all functions below like `writeBinaryAndReportMappings`

- pro: easier code
- pro: the runtime error opening the stream ( `unit -> Choice<Stream,Diagnostic>` ) stop fsc **before** compiling
- cons: the factory ( `unit -> Choice<Stream,Diagnostic>` is called always, but not deferred. The stream is open the whole compilation (not always a bad thing)

@dsyme maybe i can open a discussion in FCS about the api? This is an internal implementation detail, but the problem exists if a public api is exposed, so maybe is better to use same solution.